gnucash/gnome-utils/gnc-gtk-utils.hpp

```diff
@@ -0,0 +1,130 @@
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+\********************************************************************/
+
+#ifndef GNC_GTK_UTILS_HPP
+#define GNC_GTK_UTILS_HPP
+
+#include <string>
+#include <optional>
+
+class GtkIteratorStruct
+{
+public:
+    GtkIteratorStruct (GtkTreeModel* model, GtkTreeIter iter) : m_model{model}, m_iter{iter} {};
+
+    template <typename T>
+    T get (int column)
+    {
+        gpointer rv;
+        gtk_tree_model_get(m_model, &m_iter, column, &rv, -1);
+        return static_cast<T>(rv);
+    }
+
+    GtkTreeIter* get_iter_ptr () { return &m_iter; };
+
+    std::string get_string (int column)
+    {
+        auto str = get<char*>(column);
+        std::string rv{str};
+        g_free (str);
+        return rv;
+    }
+
+private:
+    GtkTreeModel* m_model;
+    GtkTreeIter m_iter;
+};
+
+// Custom iterator class
+template <typename ModelType>
+class GtkTreeModelIterator
+{
+public:
+    GtkTreeModelIterator(GtkTreeModel* model, std::optional<GtkTreeIter> iter) : m_model(model), m_iter(iter) {}
+
+    GtkTreeModelIterator& operator++()
+    {
+        if (!m_iter.has_value())
+            throw "no value, cannot increment";
+        if (!gtk_tree_model_iter_next(m_model, &m_iter.value()))
+            m_iter = std::nullopt;
+        return *this;
+    }
+
+    ModelType operator*() const
+    {
+        if (!m_iter.has_value())
+            throw "no value, cannot dereference";
+        return ModelType (m_model, *m_iter);
+    }
+
+    bool has_value() { return m_iter.has_value(); };
+
+    bool operator==(const GtkTreeModelIterator& other) const
+    {
+        if (m_model != other.m_model)
+            return false;
+        if (!m_iter.has_value() && !other.m_iter.has_value())
+            return true;
+        if (!m_iter.has_value() || !other.m_iter.has_value())
+            return false;
+        return (m_iter->stamp == other.m_iter->stamp) &&
+            (m_iter->user_data == other.m_iter->user_data) &&
+            (m_iter->user_data2 == other.m_iter->user_data2) &&
+            (m_iter->user_data3 == other.m_iter->user_data3);
+    }
+
+    bool operator!=(const GtkTreeModelIterator& other) const { return !(*this == other); }
+
+private:
+    GtkTreeModel* m_model;
+    std::optional<GtkTreeIter> m_iter;
+};
+
+// Custom container class
+template <typename ModelType>
+class GtkTreeModelContainer
+{
+public:
+    using TreeModelIterator = GtkTreeModelIterator<ModelType>;
+
+    GtkTreeModelContainer(GtkTreeModel* model)
+        : m_model{model}
+        , m_begin_iter{TreeModelIterator(m_model, std::nullopt)}
+        , m_end_iter{TreeModelIterator(m_model, std::nullopt)}
+    {
+        GtkTreeIter iter;
+        gtk_tree_model_get_iter_first(m_model, &iter);
+        m_begin_iter = TreeModelIterator(m_model, iter);
+    }
```

Comment:
If this last commit is desirable, this ctor will init m_begin_iter with a dummy value before finding the proper beginning iterator. If we remove the initializer line then compilation fails.